api/clients/v2/eigenda_client.go

```diff
@@ -0,0 +1,285 @@
+	}
+
+	return NewEigenDAClient(
+		log,
+		rand.New(rand.NewSource(rand.Int63())),
+		clientConfig,
+		relayClient,
+		blobVerifier,
+		codec,
+		g1Srs)
+}
+
+// NewEigenDAClient assembles an EigenDAClient from subcomponents that have already been constructed and initialized.
+func NewEigenDAClient(
+	log logging.Logger,
+	random *rand.Rand,
+	clientConfig *EigenDAClientConfig,
+	relayClient RelayClient,
+	blobVerifier verification.IBlobVerifier,
+	codec codecs.BlobCodec,
+	g1Srs []bn254.G1Affine) (*EigenDAClient, error) {
+
+	return &EigenDAClient{
+		log:          log,
+		random:       random,
+		clientConfig: clientConfig,
+		codec:        codec,
+		relayClient:  relayClient,
+		blobVerifier: blobVerifier,
+		g1Srs:        g1Srs,
+	}, nil
+}
+
+// GetPayload iteratively attempts to fetch a given blob with key blobKey from relays that have it, as claimed by the
+// blob certificate. The relays are attempted in random order.
+//
+// If the blob is successfully retrieved, then the blob is verified. If the verification succeeds, the blob is decoded
+// to yield the payload (the original user data), and the payload is returned.
+func (c *EigenDAClient) GetPayload(
+	ctx context.Context,
+	blobKey core.BlobKey,
+	eigenDACert *verification.EigenDACert) ([]byte, error) {
+
+	relayKeys := eigenDACert.BlobVerificationProof.BlobCertificate.RelayKeys
+	relayKeyCount := len(relayKeys)
+
+	blobCommitmentProto := contractEigenDABlobVerifier.BlobCommitmentBindingToProto(
+		&eigenDACert.BlobVerificationProof.BlobCertificate.BlobHeader.Commitment)
+	blobCommitment, err := encoding.BlobCommitmentsFromProtobuf(blobCommitmentProto)
+
+	if err != nil {
+		return nil, fmt.Errorf("blob commitments from protobuf: %w", err)
+	}
+
+	if relayKeyCount == 0 {
+		return nil, errors.New("relay key count is zero")
+	}
+
+	// create a randomized array of indices, so that it isn't always the first relay in the list which gets hit
+	indices := c.random.Perm(relayKeyCount)
+
+	// TODO (litt3): consider creating a utility which can deprioritize relays that fail to respond (or respond maliciously)
+
+	// iterate over relays in random order, until we are able to get the blob from someone
+	for _, val := range indices {
+		relayKey := relayKeys[val]
+
+		blob, err := c.getBlobWithTimeout(ctx, relayKey, blobKey)
+		// if GetBlob returned an error, try calling a different relay
+		if err != nil {
+			c.log.Warn("blob couldn't be retrieved from relay", "blobKey", blobKey, "relayKey", relayKey, "error", err)
+			continue
+		}
+
+		err = c.verifyBlobFromRelay(blobKey, relayKey, blob, blobCommitment.Commitment, blobCommitment.Length)
+
+		// An honest relay should never send a blob which doesn't verify
+		if err != nil {
+			c.log.Warn("verify blob from relay: %w", err)
+			continue
+		}
+
+		err = c.verifyBlobV2WithTimeout(ctx, eigenDACert)
+		if err != nil {
+			c.log.Warn("verifyBlobV2 failed", "blobKey", blobKey, "relayKey", relayKey, "error", err)
+			continue
+		}
```

Comment:
These 2 functions are a bit confusing, unclear how they are different just from the names. We basically need to verify 2 things: 1) cert is valid (thats the function call which has a weird name because its not actually verifying the blob) and 2) very the blob against the cert.

Wdyt about changing the function names to be:
- verifyBlobV2WithTimeout -> verifyCertV2WithTimeout
- verifyBlobFromRelay -> verifyBlobAgainstCert

Also I would change their order, I feel like its more natural to verify the cert first. Actually we should even verify the cert at the very top of this function, even before retrievinb the blob from the relay (b/c if the cert is invalid there's no point in retrieving the blob)

Reply:
Agree with everything, except I think `verifyCertWithTimeout` is better than `verifyCertV2WithTimeout`, since there isn't a concept of a `V2Cert` right now, and I think it's better to avoid introducing one.

Reply:
Actually, now that I think about it...

should `GetPayload` even be calling `verifyBlobV2` in the first place? Verifying the cert on-chain is totally unrelated to getting the payload from a relay. It seems like the right flow would be:

- proxy gets a DA cert
- proxy verifies the cert with `verifyBlobV2`
- If verification succeeds, proxy calls `GetPayload`

Reply:
(in the meantime, I've made the naming and ordering changes requested [ef3944db](https://github.com/Layr-Labs/eigenda/pull/972/commits/ef3944db8cf6596ee0a4c20f3d34328ed8920f10))

Reply:
Thanks for the changes! Re your architctural question, I think that could make sense. Would this work with the recent division of eigenda_client into a write and read client? Let's discuss in our sync this afternoon.

Reply:
Concerning the architectural question, let's leave the cert validation where it is for now. We could decide to move it in the future.